File: BaekjoonAlgorithm/BA_10282/BJAlgo_10282.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class BJAlgo_10282 {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int test = Integer.parseInt(br.readLine());
        for (int i = 0; i < test; i++) {
            int[] ndc = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

            List<ArrayList<int[]>> computers = new ArrayList<>();
            for (int n = 0; n < ndc[0] + 1; n++) {
                computers.add(new ArrayList<>());
            }

            for (int j = 0; j < ndc[1]; j++) {
                int[] abs = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
                computers.get(abs[1]).add(new int[]{abs[0], abs[2]});
            }

            int[] result = dijkstra(computers, ndc[2]);
            int count = 0;
            int max_time = -1;
            for (int r : result) {
                if (r != 100000001) {
                    count += 1;
                    if (max_time < r) {
                        max_time = r;
                    }
                }
            }

            System.out.println(count + " " + max_time);
        }
    }

    private static int[] dijkstra(List<ArrayList<int[]>> computers, int start) {
        int[] time = new int[computers.size()];
        Arrays.fill(time, 100000001);
        time[start] = 0;

        PriorityQueue<int[]> queue = new PriorityQueue<>((int[] c1, int[] c2) -> c1[0] <= c2[0] ? 1 : -1);
        queue.offer(new int[]{0, start});

        while (!queue.isEmpty()) {
            int[] current = queue.poll();  // 0: current_time, 1: current_node

            if (current[0] > time[current[1]]) {
                continue;
            }

            for (int[] adj : computers.get(current[1])) {  // 0: adj, 1: time
                int dist = adj[1] + current[0];
                if (time[adj[0]] > dist) {
                    time[adj[0]] = dist;
                    queue.offer(new int[]{dist, adj[0]});
                }
            }
        }

        return time;
    }
}

File: BaekjoonAlgorithm/202012/BA_10282/BJAlgo_10282.py
from sys import stdin
import heapq


def dijkstra(comp, start):
    time = [float('inf')] * len(comp)
    time[start] = 0

    queue = []
    heapq.heappush(queue, [0, start])

    while queue:
        current_time, current_node = heapq.heappop(queue)

        if current_time > time[current_node]:
            continue

        for adj, t in comp[current_node]:
            dist = current_time + t
            if dist < time[adj]:
                time[adj] = dist
                heapq.heappush(queue, [dist, adj])

    return time


test_case = int(stdin.readline())
for _ in range(test_case):
    n, d, c = map(int, stdin.readline().split())
    computers = [[] for _ in range(n + 1)]
    for _ in range(d):
        a, b, s = map(int, stdin.readline().split())
        computers[b].append([a, s])
    result = dijkstra(computers, c)
    count = 0
    max_time = -1
    for r in result:
        if r != float('inf'):
            count += 1
            if max_time < r:
                max_time = r

    print(count, max_time)
